### PR TITLE
Feature/logout

### PR DESCRIPTION
**Pull request by cicnavi**
Add RP-Initiated Logout and Back-Channel Logout functionality.

### CODE REVIEW

File: lib/Services/JsonWebTokenBuilderService.php [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+use SimpleSAML\Module\oidc\Utils\FingerprintGenerator;
+use SimpleSAML\Module\oidc\Utils\UniqueIdentifierGenerator;
+
+class JsonWebTokenBuilderService
+{
+    protected ConfigurationService $configurationService;
+    protected Configuration $jwtConfig;
+
+    /**
+     * @throws ReflectionException
```
Comment (pradtke) [inline lib/Services/JsonWebTokenBuilderService.php:24]: missing @param for the constructor argument

File: lib/Services/JsonWebTokenBuilderService.php [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+     * @throws Exception
+     */
+    public function __construct(
+        ?ConfigurationService $configurationService = null
+    ) {
+        $this->configurationService = $configurationService ?? new ConfigurationService();
```
Comment (pradtke) [inline lib/Services/JsonWebTokenBuilderService.php:30]: I don't understand why ConfigurationService is optional. Isn't this instantiated in the Container and so the dependency should be available?

File: lib/Controller/LogoutController.php [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+                            $exception->getMessage()
+                        )
+                    );
+                }
+            }
+
+            $sessionLogoutTicketStore->deleteMultiple(array_map(fn($slt) => $slt['sid'], $sessionLogoutTickets));
+        }
+
+        (new BackChannelLogoutHandler())->handle($relyingPartyAssociations);
```
Comment (pradtke) [inline lib/Controller/LogoutController.php:184]: Should this come from the Container/constructor injection?

File: lib/Server/LogoutHandlers/BackChannelLogoutHandler.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+{
+    protected LogoutTokenBuilder $logoutTokenBuilder;
+
+    protected LoggerService $loggerService;
+
+    public function __construct(
+        ?LogoutTokenBuilder $logoutTokenBuilder = null,
+        ?LoggerService $loggerService = null
+    ) {
+        $this->logoutTokenBuilder = $logoutTokenBuilder ?? new LogoutTokenBuilder();
```
Comment (pradtke) [inline lib/Server/LogoutHandlers/BackChannelLogoutHandler.php:28]: Similar to other places : I'm not sure why the object isn't instantiated in the Container and with required constructor args.

**Review by pradtke — COMMENTED** [review]
@cicnavi wow!  the logout spec is way more complicated than I had expected.  Well done on such a huge contribution.  My comments are mostly related to allowing null constructor arguments, and instead of always creating objects in the `Container` class and just using those.

File: lib/Services/JsonWebTokenBuilderService.php (hunk 1 above)
Comment (cicnavi) [reply lib/Services/JsonWebTokenBuilderService.php:24]: ok

**Comment by cicnavi** [comment]
Reasons for optional arguments and currently not using Container:

Container property \SimpleSAML\Module\oidc\Services\Container::$services is (currently) not static, meaning it will have to be initialized on every Container constructor call.

\SimpleSAML\Module\oidc\Controller\LogoutController::logoutHandler is a static method. Note that this method can be called without LogoutHandler ever being constructed (for example from SSP). 

In this static LogoutController::logoutHandler() method I have to use \SimpleSAML\Module\oidc\Server\LogoutHandlers\BackChannelLogoutHandler, which internally uses other mentioned services.

If I were to use (new Container())->get('BackChannelLogoutHandler), I would trigger huge Container constructor again...

I would use Container, but only if I avoided running its huge constructor multiple times. So, I'm open to refactor this if you agree to convert Container::$services property to static one, or if you have any other suggestion...

File: lib/Services/JsonWebTokenBuilderService.php (hunk 2 above)
Comment (cicnavi) [reply lib/Services/JsonWebTokenBuilderService.php:30]: Check comment.

File: lib/Controller/LogoutController.php (hunk 3 above)
Comment (cicnavi) [reply lib/Controller/LogoutController.php:184]: Check comment.

File: lib/Server/LogoutHandlers/BackChannelLogoutHandler.php (hunk 4 above)
Comment (cicnavi) [reply lib/Server/LogoutHandlers/BackChannelLogoutHandler.php:28]: Check comment.

**Comment by pradtke** [comment]
Thanks for the explanation.  
I don't think we should make the Container $services static, since that ends up making testing a pain (with state carrying over between tests).   Some dependency management tools support lazy creation of the objects, but ours does not and I don't think it's worth the effort to change that.

Perhaps move all `new`s/object creation into the static LogoutHandler, and then those classes will just required arguments. e.g. create JsonWebTokenBuilderService, LogoutTokenBuilder, etc all in the the LogoutHandler and pass them via the constructors?    In the end, I think it is ok to merge with the way you have it.

**Comment by cicnavi** [comment]
In general I agree, but can't say it would necessarily make testing a pain in this particular case, since we practically don't use Container in tests. Other classes are inject-able with dependencies, so we easily use mocks or dependency instances... For the container itself we could introduce setter method to (re)set service... However, I don't really mind, since we are to at some point in the future move to SSP Symfony container...

If you agree, I would leave it like this. I don't feel comfortable creating all those instances in static handler method, seems ugly to me... (configuration instance for JsonService, then JsonService for TokenBuiler, then TokenBuilder and LoggerService for LogoutHandler and then Handler itself... ). I don't see any downside of this approach at the moment (services are still inject-able to those classes).

**Pull request merged**